db/compaction/compaction_job.cc

```diff
@@ -945,6 +950,10 @@ void CompactionJob::ProcessKeyValueCompaction(SubcompactionState* sub_compact) {
         key, value, ikey.sequence, ikey.type);
     sub_compact->num_output_records++;
 
+    if (partitioner.get() != nullptr) {
+      partitioner->Reset(key);
```

Comment:
Can we merge the `Reset` and `ShouldPartition` call into one method call? (echoing the same comment from https://github.com/facebook/rocksdb/pull/5201#discussion_r278659718)

Reply:
It is possible, but then the loop will need to copy previous key. Right now the partitioner can remember only the bit it needs. It is hard to say if that copy will take more time, but I think virtual call is pretty cheap compared to memory copy of X number of bytes. So should I do that?

Reply:
good point about memcopy. But it looks like in the current approach the `Reset` needs a memcopy as well. The `key` is a slice obtain from `c_iter`, which lifetime is ended after `c_iter->Next()` is called. The partitioner needs to make memcopy to keep it. I'm not sure why ASAN test don't complain. 

In case I didn't misread the code and memcopy is inevitable, I think we can change to the following flow:
```
for KV in c_iter {
    if (partitioner->ShouldPartition(next_key)) {
         output_file_ended = true;
    }
    if (output_file_ended) {
         partitioner->Reset(next_key);
    }
}
``` 
That way the partitioner is invoke (#keys + #files) times, and it can decide on its own whether to memcopy keys per key or per file.

Reply:
reset removed and should partition is with two arguments